### PR TITLE
MueLu: regionMG add a region matvec to xpetra

### PR DESCRIPTION
**Pull request by pohm01**
@trilinos/muelu 

## Motivation
For use with the region mg. This is a first step towards further integration of the region mg method into MueLu, with the future goal of allowing the region mg to use preexisting functions and packages.

## Testing
MueLu unit tests and region mg tests pass on my local machine.
The region matvec is tested in the region matrix unit tests.

### CODE REVIEW

File: packages/xpetra/src/CrsMatrix/Xpetra_CrsMatrix.hpp [hunk 1]
```diff
@@ -260,6 +260,9 @@ namespace Xpetra {
      */
     virtual void apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode=Teuchos::NO_TRANS, Scalar alpha=ScalarTraits< Scalar >::one(), Scalar beta=ScalarTraits< Scalar >::zero()) const = 0;
 
+  virtual void apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta, bool sumInterfaceValues, const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs) const = 0;
```
Comment (mayrmt) [inline packages/xpetra/src/CrsMatrix/Xpetra_CrsMatrix.hpp:263]: Indentation:
```suggestion
    virtual void apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta, bool sumInterfaceValues, const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs) const = 0;
```

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp [hunk 2]
```diff
@@ -335,6 +336,30 @@ namespace Xpetra {
...
 
+    template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+    void TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>::apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta, bool sumInterfaceValues, const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs ) const
+    {
+        RCP<const Map< LocalOrdinal, GlobalOrdinal, Node >> regionInterfaceMap = regionInterfaceImporter->getTargetMap();
+        XPETRA_MONITOR("TpetraCrsMatrix::apply(region)");
+        mtx_->localApply(toTpetra(X), toTpetra(Y), mode, alpha, beta);
+        if (sumInterfaceValues)
+        {
+          // Step 2: preform communication to propagate local interface
```
Comment (mayrmt) [inline packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:347]: There's no "Step 1" before this "Step 2". Same for "Step 3" below.

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp [hunk 3]
```diff
@@ -335,6 +336,30 @@ namespace Xpetra {
...
+          // values to all the processor that share interfaces.
+          RCP<MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node >> matvecInterfaceTmp = MultiVectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(regionInterfaceMap, 1);
```
Comment (mayrmt) [inline packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: Out of curiosity: Can't you use the pre-defined short names here to improve readability?
```suggestion
          RCP<MultiVector> matvecInterfaceTmp = MultiVectorFactory::Build(regionInterfaceMap, 1);
```

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp [hunk 4]
```diff
@@ -335,6 +336,30 @@ namespace Xpetra {
...
+          matvecInterfaceTmp->doImport(Y, *regionInterfaceImporter, INSERT);
+
+          // Step 3: sum all contributions to interface values
+          // on all ranks
+          //ArrayRCP<Scalar> YData = Y->getDataNonConst(0);
```
Comment (mayrmt) [inline packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:354]: Please remove commented code.
```suggestion
```

**Review by mayrmt — CHANGES_REQUESTED** [review]
@pohm01 Thanks for pushing this forward! I made some minor remarks throughout the code. And I have some more general remarks/questions that are applicable to more than one place in the code, so I collect them here in this comment:

### Testing

Please make sure that your branch contains the changes from PR #8506, where we have fixed some problems in the region MG regression tests. Only with the improved testing procedure, we can be sure that everything works as expected. The fix is part of `develop` already, so please just rebase if necessary. Thanks!

### Documentation

Please add some documentation to explain the use case of this method and also the non-traditional function arguments related to  the region interfaces. This is particularly important, since these methods are added to the regular Xpetra matrix classes (not to specialized region classes). We really need to make sure to

1. help a user to pick the correct `apply()` routine by distinguishing between the "normal" and the "region" scenario.
2. document the necessary region information to be provided by the user (which isn't well documented at the moment).

### `const` qualifiers

Wouldn't it make more sense for the last two arguments to move the `const` qualifier from `const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs` to `RCP<const Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, Teuchos::ArrayRCP<const LocalOrdinal>& regionInterfaceLIDs`? This is data provided by the user and it should not be modifiable within the `apply()` routine, so the `const` keyword should protect the data, not the pointers.

Maybe also add `const` to the scalar parameters `alpha` and `beta`.

### `ArrayRCP` vs. `ArrayView`

I'm wondering (and curious about other opinions) if the last argument could or should be a `Teuchos::ArrayView<const LocalOrdinal>& regionInterfaceLIDs`. Maybe @lucbv has some perspective on that.

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp (hunk 3 above)
Comment (pohm01) [reply packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: The short names don't appear to be defined yet for this file. Should I add the shortnames header?

Comment (mayrmt) [reply packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: @pohm01 Well, I think you can just include existing Xpetra's short name headers, right?

File: packages/xpetra/src/CrsMatrix/Xpetra_CrsMatrix.hpp [hunk 5]
```diff
@@ -260,6 +260,10 @@ namespace Xpetra {
      */
     virtual void apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode=Teuchos::NO_TRANS, Scalar alpha=ScalarTraits< Scalar >::one(), Scalar beta=ScalarTraits< Scalar >::zero()) const = 0;
 
+    //! matrix-multivector multiplication for region layout matrices
+    virtual void apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta, bool sumInterfaceValues, const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs) const = 0;
```
Comment (mayrmt) [inline packages/xpetra/src/CrsMatrix/Xpetra_CrsMatrix.hpp:264]: @pohm01 Please make sure to document `regionInterfaceImporter` and `regionInterfaceLIDs`. We need some form of description, such that a user knows, what he/she has to provide.

**Review by mayrmt — COMMENTED** [review]
@pohm01 I've made one more comment regarding documentation. Other than that, changes look good to me. Thanks!

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp [hunk 6]
```diff
@@ -335,6 +336,29 @@ namespace Xpetra {
     template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
     void TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>::apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta) const { XPETRA_MONITOR("TpetraCrsMatrix::apply"); mtx_->apply(toTpetra(X), toTpetra(Y), mode, alpha, beta); }
 
+    template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+    void TpetraCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node>::apply(const MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &X, MultiVector< Scalar, LocalOrdinal, GlobalOrdinal, Node > &Y, Teuchos::ETransp mode, Scalar alpha, Scalar beta, bool sumInterfaceValues, const RCP<Import<LocalOrdinal, GlobalOrdinal, Node> >& regionInterfaceImporter, const Teuchos::ArrayRCP<LocalOrdinal>& regionInterfaceLIDs ) const
+    {
+        RCP<const Map< LocalOrdinal, GlobalOrdinal, Node >> regionInterfaceMap = regionInterfaceImporter->getTargetMap();
+        XPETRA_MONITOR("TpetraCrsMatrix::apply(region)");
```
Comment (lucbv) [inline packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:343]: Actually maybe move that up one line so everything is inside the timing region.

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp (hunk 3 above)
Comment (lucbv) [reply packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: I am not sure I see where the change has been made but the idea is to include this header: `Xpetra_UseShortNamesScalar.hpp`

**Review by lucbv — COMMENTED** [review]
This looks fine, thanks for the performance improvements!

File: packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp (hunk 3 above)
Comment (mayrmt) [reply packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: I don't think, that these changes have been made yet. Can't see them either.

Comment (pohm01) [reply packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp:349]: I haven't made the changes for this yet. Where do you suggest I put shortnames header? The only place I have found that does not cause problems is inside the function definition here.

**Review by mayrmt — DISMISSED** [review]
@pohm Thanks for the final modifications. Feel free to merge!

**Review by mayrmt — DISMISSED** [review]

**Review by mayrmt — APPROVED** [review]

**Pull request merged**